src/gluonts/mx/trainer/_base.py

```diff
@@ -204,250 +244,230 @@ def __call__(
-
-                    batch_iter = itertools.islice(
-                        batch_iter, num_batches_to_use
+                if is_training:
+                    self.callbacks.on_train_batch_start(training_network=net)
+                else:
+                    self.callbacks.on_validation_batch_start(
+                        training_network=net
                     )
 
-                    with tqdm(batch_iter, total=num_batches_to_use) as it:
-                        for batch_no, batch in enumerate(it, start=1):
-                            # `batch` here is expected to be a dictionary whose fields
-                            # should correspond 1-to-1 with the network inputs
-                            # see below how `batch.values()` is fed into the network
-
-                            if first_forward:
-                                first_forward = False
-                                _ = net(*batch.values())
-                                if self.post_initialize_cb:
-                                    self.post_initialize_cb(net)
-
-                            with mx.autograd.record():
-                                # we set the mode explicitly as by default mxnet assumes predict mode and hence
-                                # dropout layers are not used if the mode is not explicitly set to training
-                                mode = (
-                                    autograd.train_mode
-                                    if is_training
-                                    else autograd.predict_mode
-                                )
-                                with mode():
-                                    output = net(*batch.values())
-
-                                # network can returns several outputs, the first being always the loss
-                                # when having multiple outputs, the forward returns a list in the case of hybrid and a
-                                # tuple otherwise
-                                # we may wrap network outputs in the future to avoid this type check
-                                if isinstance(output, (list, tuple)):
-                                    loss = output[0]
-                                else:
-                                    loss = output
-
-                                batch_size = loss.shape[0]
-
-                            if not np.isfinite(ndarray.sum(loss).asscalar()):
-                                logger.warning(
-                                    "Batch [%d] of Epoch[%d] gave NaN loss and it will be ignored",
-                                    batch_no,
-                                    epoch_no,
-                                )
+                batch_iter = itertools.islice(batch_iter, num_batches_to_use)
+
+                with tqdm(batch_iter, total=num_batches_to_use) as it:
+                    for batch_no, batch in enumerate(it, start=1):
```

Comment:
Maybe we can avoid one level of indentation here and avoid the `with` clause: just `it = ...` should work